src/gui/editors/PianoRoll.cpp

```diff
@@ -2856,208 +2869,256 @@ void PianoRoll::paintEvent(QPaintEvent * pe )
-			// WHITE_KEY_SMALL_HEIGHT, so WHITE_KEY_SMALL_HEIGHT/2
-			// is smaller)
-			first_white_key_height = WHITE_KEY_SMALL_HEIGHT / 2;
+			m_notesEditHeight = m_oldNotesEditHeight;
 		}
-		// check whether to draw a big or a small white key
-		if( prKeyOrder[key % KeysPerOctave] == PR_WHITE_KEY_SMALL )
+		pianoArea = keyAreaBottom() - keyAreaTop();
+		m_pianoKeysVisible = pianoArea / KEY_LINE_HEIGHT;
+		partialKeyVisible = pianoArea % KEY_LINE_HEIGHT;
+		if (m_pianoKeysVisible >= NumKeys)
 		{
-			// draw a small one while checking if it is pressed or not
-			if( hasValidPattern() && m_pattern->instrumentTrack()->pianoModel()->isKeyPressed( key ) )
-			{
-				p.drawPixmap( PIANO_X, y - WHITE_KEY_SMALL_HEIGHT, *s_whiteKeySmallPressedPm );
-			}
-			else
-			{
-				p.drawPixmap( PIANO_X, y - WHITE_KEY_SMALL_HEIGHT, *s_whiteKeySmallPm );
-			}
-			// update y-pos
-			y -= WHITE_KEY_SMALL_HEIGHT;
+			// always increase m_notesEditHeight to compensate
+			m_pianoKeysVisible = NumKeys;
+			m_notesEditHeight = height() - (NumKeys * KEY_LINE_HEIGHT) -
+				PR_TOP_MARGIN - PR_BOTTOM_MARGIN;
+			partialKeyVisible = 0;
+		}
+		topKey = qBound(0, m_startKey + m_pianoKeysVisible - 1, NumKeys - 1);
+		topNote = topKey % KeysPerOctave;
+		// if resizing the note edit area, don't touch m_oldNotesEditHeight
+		if (m_action != ActionResizeNoteEditArea && partialKeyVisible != 0)
+		{
+			m_oldNotesEditHeight = m_notesEditHeight;
+			// adding height always means we don't have to add keys
+			m_notesEditHeight += partialKeyVisible;
+		}
+		updatePositionLineHeight();
+		int x, q = quantization(), tick;
 
+		// draw vertical quantization lines
+		// If we're over 100% zoom, we allow all quantization level grids
+		if (m_zoomingModel.value() <= 3)
+		{
+			// we're under 100% zoom
+			// allow quantization grid up to 1/24 for triplets
+			if (q % 3 != 0 && q < 8) { q = 8; }
+			// allow quantization grid up to 1/32 for normal notes
+			else if (q < 6) { q = 6; }
 		}
-		else if( prKeyOrder[key % KeysPerOctave] == PR_WHITE_KEY_BIG )
+		auto xCoordOfTick = [=](int tick) {
+			return WHITE_KEY_WIDTH + (
+				(tick - m_currentPosition) * m_ppt / MidiTime::ticksPerTact()
+			);
+		};
+		p.setPen(lineColor());
+		for (tick = m_currentPosition - m_currentPosition % q,
+			x = xCoordOfTick(tick);
+			x <= width();
+			tick += q, x = xCoordOfTick(tick))
 		{
-			// draw a big one while checking if it is pressed or not
-			if( hasValidPattern() && m_pattern->instrumentTrack()->pianoModel()->isKeyPressed( key ) )
+			p.drawLine(x, keyAreaTop(), x, noteEditBottom());
+		}
+
+		// draw horizontal grid lines and piano notes
+		p.setClipRect(0, keyAreaTop(), width(), keyAreaBottom() - keyAreaTop());
+		int key_line_y = keyAreaTop();
+		int grid_line_y = keyAreaTop() + KEY_LINE_HEIGHT - 1;
+		static const int yTextCorrections[KeysPerOctave] = {
+		/*   C      D      E   F      G      A      B   */
+			-4, 0, -8, 0, -6, -4, 0, -8, 0, -8, 0, -6
```

Comment:
This looks right with 4-space-tabs, but in case someone uses invalid tabs, I'd rather change this:

```suggestion
		    -4, 0, -8, 0, -6, -4, 0, -8, 0, -8, 0, -6
```